azurerm/internal/services/subscription/subscription_resource.go

```diff
@@ -0,0 +1,472 @@
+		}
+
+		req.Properties.SubscriptionID = utils.String(subscriptionId)
+		existingSub, err := client.Get(ctx, subscriptionId)
+		if err != nil {
+			return fmt.Errorf("could not read existing Subscription %q", subscriptionId)
+		}
+		// Disabled and Warned are both "effectively" cancelled states,
+		if existingSub.State == subscriptions.Disabled || existingSub.State == subscriptions.Warned {
+			log.Printf("[DEBUG] Existing subscription in Disabled/Cancelled state Terraform will attempt to re-activate it")
+			if _, err := subscriptionClient.Enable(ctx, subscriptionId); err != nil {
+				return fmt.Errorf("enabling Subscription %q: %+v", subscriptionId, err)
+			}
+			deadline, _ := ctx.Deadline()
+			createDeadline := time.Until(deadline)
+			if err := waitForSubscriptionStateToSettle(ctx, meta.(*clients.Client), subscriptionId, "Active", createDeadline); err != nil {
+				return fmt.Errorf("failed waiting for Subscription %q (Alias %q) to enter %q state: %+v", subscriptionId, id.Name, "Active", err)
+			}
+
+		}
+	} else {
+		// If we're not assuming control of an existing Subscription, we need to know where to create it.
+		req.Properties.DisplayName = utils.String(d.Get("subscription_name").(string))
+		req.Properties.BillingScope = utils.String(d.Get("billing_scope_id").(string))
+	}
+
+	future, err := aliasClient.Create(ctx, aliasName, req)
+	if err != nil {
+		return fmt.Errorf("creating new Subscription (Alias %q): %+v", aliasName, err)
+	}
+
+	if err := future.WaitForCompletionRef(ctx, aliasClient.Client); err != nil {
+		return fmt.Errorf("waiting for creation of Subscription with Alias %q: %+v", id.Name, err)
+	}
+
+	alias, err := aliasClient.Get(ctx, id.Name)
+	if err != nil || alias.Properties == nil || alias.Properties.SubscriptionID == nil {
+		return fmt.Errorf("failed reading subscription details for Alias %q: %+v", id.Name, err)
+	}
+
+	deadline, _ := ctx.Deadline()
+	createDeadline := time.Until(deadline)
+
+	if err := waitForSubscriptionStateToSettle(ctx, meta.(*clients.Client), *alias.Properties.SubscriptionID, "Active", createDeadline); err != nil {
+		return fmt.Errorf("failed waiting for Subscription %q (Alias %q) to enter %q state: %+v", subscriptionId, id.Name, "Active", err)
+	}
+
+	d.SetId(id.ID())
+
+	return resourceSubscriptionRead(d, meta)
+}
+
+func resourceSubscriptionUpdate(d *schema.ResourceData, meta interface{}) error {
+	aliasClient := meta.(*clients.Client).Subscription.AliasClient
+	subscriptionClient := meta.(*clients.Client).Subscription.SubscriptionClient
+	ctx, cancel := timeouts.ForUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := parse.SubscriptionAliasID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	locks.ByName(id.Name, SubscriptionResourceName)
+	defer locks.UnlockByName(id.Name, SubscriptionResourceName)
```

Comment:
do we want to lock on the Subscription ID here instead?

Reply:
There are effectively 2 resources at play here, so I've (defensively) opted to try and guard against change to either, since we need to maintain the 1:1 relation.